### PR TITLE
No bug - Rewrite the context menu code.

### PR DESCRIPTION
**Pull request by emilio**
This is not for perf, just to try to make it saner.

### CODE REVIEW

File: static/js/context-menu.js [hunk 1]
```diff
@@ -1,107 +1,170 @@
...
+    return s.replace("_", data);
+  }
+
+  tryShowOnClick(event) {
     // Don't display the context menu if there's a selection.
     // User could be trying to select something and the context menu will undo it.
-    return;
-  }
+    if (!window.getSelection().isCollapsed)
+      return;
```
Comment (asutherland) [inline static/js/context-menu.js:35]: Curly.

File: static/js/context-menu.js [hunk 2]
```diff
@@ -1,107 +1,170 @@
...
 
-  var top = event.clientY + window.scrollY;
-  var left = event.clientX;
+    if (!event.target.closest("code"))
+      return;
```
Comment (asutherland) [inline static/js/context-menu.js:38]: curly

File: static/js/context-menu.js [hunk 3]
```diff
@@ -1,107 +1,170 @@
...
+      let visibleToken = token.textContent;
+      menuItems.push({html: "Sticky highlight",
+                      href: `javascript:Hover.stickyHighlight('${symbols}', '${visibleToken}')`});
+    }
 
-window.addEventListener("pageshow", function() {
-  $('#context-menu').remove();
-}, false);
+    if (!menuItems.length)
+      return;
```
Comment (asutherland) [inline static/js/context-menu.js:80]: curly

File: static/js/context-menu.js [hunk 4]
```diff
@@ -1,107 +1,170 @@
...
+  }
+};
+
+var Hover = new class Hover {
+  constructor() {
+    this.items = [];
+    this.sticky = false;
+    window.addEventListener('mousedown', () => {
+      if (this.sticky)
+        this.deactivate()
```
Comment (asutherland) [inline static/js/context-menu.js:109]: curly

File: static/js/context-menu.js [hunk 5]
```diff
@@ -1,107 +1,170 @@
...
+  }
 
-  var elt = document.elementFromPoint(x, y);
-  while (!elt.hasAttribute("data-symbols")) {
-    elt = elt.parentNode;
-    if (!elt || !(elt instanceof Element)) {
-      hovered.removeClass("hovered");
-      hovered = $();
+  _handleMouseMove(event) {
+    if (ContextMenu.active || this.sticky)
```
Comment (asutherland) [inline static/js/context-menu.js:116]: curly

File: static/js/context-menu.js [hunk 6]
```diff
@@ -1,107 +1,170 @@
...
       return;
-    }
+
+    let symbols = event.target.closest("[data-symbols]");
+    if (!symbols)
+      return this.deactivate();
```
Comment (asutherland) [inline static/js/context-menu.js:121]: curly

File: static/js/context-menu.js [hunk 7]
```diff
@@ -1,107 +1,170 @@
...
+
+    this.activate(symbols.getAttribute("data-symbols"), symbols.textContent);
+  }
+
+  deactivate() {
+    for (let item of this.items)
+      item.classList.remove("hovered");
```
Comment (asutherland) [inline static/js/context-menu.js:128]: Definitely curly.

File: static/js/context-menu.js [hunk 8]
```diff
@@ -1,107 +1,170 @@
...
 
-  hovered.removeClass("hovered");
-  hovered = findReferences(elt.getAttribute("data-symbols"), elt.textContent);
-  hovered.addClass("hovered");
-});
+  activate(symbols, visibleToken) {
+    this.deactivate();
+    this.items = this.findReferences(symbols, visibleToken);
+    for (let item of this.items)
+      item.classList.add("hovered");
```
Comment (asutherland) [inline static/js/context-menu.js:136]: Definitely curly.

File: static/js/context-menu.js [hunk 9]
```diff
@@ -1,107 +1,170 @@
...
+  }
 
-function stickyHighlight(symbols, visibleToken)
-{
-  $('#context-menu').remove();
+  findReferences(symbols, visibleToken) {
+    function symbolsFromString(symbols) {
+      if (!symbols || symbols == "?") // XXX why the `?` special-case?
+        return [];
```
Comment (asutherland) [inline static/js/context-menu.js:142]: curly

File: static/js/context-menu.js [hunk 10]
```diff
@@ -1,107 +1,170 @@
...
+      return symbols.split(",");
+    }
 
-  hovered.removeClass("hovered");
-  hovered = findReferences(symbols, visibleToken);
-  hovered.addClass("hovered");
+    symbols = symbolsFromString(symbols);
+    if (!symbols.length)
+      return [];
```
Comment (asutherland) [inline static/js/context-menu.js:148]: curly

File: static/js/context-menu.js [hunk 11]
```diff
@@ -1,107 +1,170 @@
...
+    ContextMenu.hide();
+    this.activate(symbols, visibleToken);
+    this.sticky = true;
+  }
+};
+
+function getTargetWord() {
   var selection = window.getSelection();
   if (!selection.isCollapsed) {
     return null;
```
Comment (asutherland) [inline static/js/context-menu.js:170]: curly

File: static/js/blame.js [hunk 12]
```diff
@@ -161,7 +161,7 @@ var BlameStripHoverHandler = new class BlameStripHoverHandler {
 
   handleEvent(event) {
     // Suppress the blame hover popup if the context menu is visible.
-    if (document.getElementById("context-menu"))
+    if (ContextMenu.active)
       return;
```
Comment (asutherland) [inline static/js/blame.js:165]: Curly.

**Comment by emilio** [comment]
alright alright. I did it out of personal preference, but 2v1, curly wins :)

**Comment by emilio** [comment]
I curlied all the things. Given this was your only objections I'll merge and open a separate PR for the eslint / prettier config.

**Pull request merged**